### ISSUE TITLE
```FlopCounterMode``` returns 0 when inference mode is on during forwardpropagation.

### ISSUE THREAD

**Issue by StepinSilence** [pytorch/pytorch#126268]
### 🐛 Describe the bug

```FlopCounterMode``` reports 0 FLOPS when the forward propagation has the inference mode enabled by ```torch.inference_mode()```. The counter works fine if I use ```torch_no_grad()``` instead. The documentation says ```torch.inference_mode()``` disables the Forward-mode AD while ```torch.no_grad()``` does not. Maybe ```FlopCounterMode``` somehow relies on Forward-mode AD to calculate the computation cost?

PyTorch: 2.3.0
Python: 3.11.9

Test script 1 uses ```torch.inference_mode()```
```python
import torchvision.models as models
import torch
from torch.utils.flop_counter import FlopCounterMode

device = 'cuda'

inp = torch.randn(1, 3, 224, 224, device = device)
mod = models.resnet18().to(device)

@torch.inference_mode(True)
def run_batch(x):
    flop_counter = FlopCounterMode(mod)
    with flop_counter:
        return mod(x).sum()

run_batch(inp)
```

Output:
```
Module      FLOP    % Total
--------  ------  ---------
Global         0         0%
```

Test script 2 uses ```torch.no_grad()```
```python
import torchvision.models as models
import torch
from torch.utils.flop_counter import FlopCounterMode

device = 'cuda'

inp = torch.randn(1, 3, 224, 224, device = device)
mod = models.resnet18().to(device)

@torch.no_grad()
def run_batch(x):
    flop_counter = FlopCounterMode(mod)
    with flop_counter:
        return mod(x).sum()

run_batch(inp)
```

Output:
```
Module                     FLOP    % Total
--------------------  ---------  ---------
ResNet                3628.147M    100.00%
 - aten.convolution   3627.123M     99.97%
 - aten.addmm            1.024M      0.03%
 ResNet.conv1          236.028M      6.51%
  - aten.convolution   236.028M      6.51%
 ResNet.layer1         924.844M     25.49%
  - aten.convolution   924.844M     25.49%
 ResNet.layer2         822.084M     22.66%
  - aten.convolution   822.084M     22.66%
 ResNet.layer3         822.084M     22.66%
  - aten.convolution   822.084M     22.66%
 ResNet.layer4         822.084M     22.66%
  - aten.convolution   822.084M     22.66%
 ResNet.fc               1.024M      0.03%
  - aten.addmm           1.024M      0.03%
```

Test script 3 keeps gradient calculation enabled.
```python
import torchvision.models as models
import torch
from torch.utils.flop_counter import FlopCounterMode

device = 'cuda'

inp = torch.randn(1, 3, 224, 224, device = device)
mod = models.resnet18().to(device)

def run_batch(x):
    flop_counter = FlopCounterMode(mod)
    with flop_counter:
        return mod(x).sum()

run_batch(inp)
```

Output:
```
Module                     FLOP    % Total
--------------------  ---------  ---------
ResNet                3628.147M    100.00%
 - aten.convolution   3627.123M     99.97%
 - aten.addmm            1.024M      0.03%
 ResNet.conv1          236.028M      6.51%
  - aten.convolution   236.028M      6.51%
 ResNet.layer1         924.844M     25.49%
  - aten.convolution   924.844M     25.49%
 ResNet.layer2         822.084M     22.66%
  - aten.convolution   822.084M     22.66%
 ResNet.layer3         822.084M     22.66%
  - aten.convolution   822.084M     22.66%
 ResNet.layer4         822.084M     22.66%
  - aten.convolution   822.084M     22.66%
 ResNet.fc               1.024M      0.03%
  - aten.addmm           1.024M      0.03%
```

### Versions

```
PyTorch version: 2.3.0
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Red Hat Enterprise Linux Server release 7.9 (Maipo) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
Clang version: 3.4.2 (tags/RELEASE_34/dot2-final)
CMake version: version 2.8.12.2
Libc version: glibc-2.17

Python version: 3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:36:13) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.118.1.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla V100-PCIE-16GB
Nvidia driver version: 550.54.15
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                28
On-line CPU(s) list:   0-27
Thread(s) per core:    2
Core(s) per socket:    14
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Gold 6132 CPU @ 2.60GHz
Stepping:              4
CPU MHz:               2949.121
CPU max MHz:           3700.0000
CPU min MHz:           1000.0000
BogoMIPS:              5200.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              19712K
NUMA node0 CPU(s):     0-27
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba rsb_ctxsw ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_pkg_req pku ospke md_clear spec_ctrl intel_stibp flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.3.0
[pip3] torchaudio==2.3.0
[pip3] torchdiffeq==0.2.2
[pip3] torchvision==0.18.0
[pip3] triton==2.3.0
[conda] blas                      1.0                         mkl    conda-forge
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libblas                   3.9.0            16_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            16_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            16_linux64_mkl    conda-forge
[conda] mkl                       2022.2.1         h84fe81f_16997    conda-forge
[conda] numpy                     1.26.4          py311h64a7726_0    conda-forge
[conda] pytorch                   2.3.0           py3.11_cuda11.8_cudnn8.7.0_0    pytorch
[conda] pytorch-cuda              11.8                 h7e8668a_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.3.0               py311_cu118    pytorch
[conda] torchdiffeq               0.2.2              pyhd8ed1ab_0    conda-forge
[conda] torchtriton               2.3.0                     py311    pytorch
[conda] torchvision               0.18.0              py311_cu118    pytorch

```

cc @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by mikaylagawarecki** [comment]
cc @Chillee

**Comment by Chillee** [comment]
This is somewhat expected. Basically, `inference_mode` disables dispatching. We could probably just warn when detecting that inference mode is on.

**Comment by albanD** [comment]
Is it expected? inference_mode only skips dispatching the autograd key. All the other keys should be called as usual (in particular the Python key).

**Comment by StepinSilence** [comment]
@Chillee Thanks for your info about this. A warning definitely helps people understand why the counter does not work. But I share the concern from @albanD given that ```inference_mode()``` seemingly is more preferred than ```no_grad()``` on model inference. I am also afraid that the Flops counting works for ```no_grad()``` but not ```inference_mode()```  may not be an expected behavior for some users.

**Comment by ezyang** [comment]
I don't think this will be too difficult to fix for a motivated user. If inference mode is preventing decomps from happening, the idea is to just manually apply the decomp when you get to float counter.

I confirm this by adding this debug print

```
diff --git a/torch/utils/flop_counter.py b/torch/utils/flop_counter.py
index d7080c9e4e3..ed284c17196 100644
--- a/torch/utils/flop_counter.py
+++ b/torch/utils/flop_counter.py
@@ -688,6 +688,7 @@ class FlopCounterMode(TorchDispatchMode):
             print(self.get_table(self.depth))
 
     def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        print(func)
         kwargs = kwargs if kwargs else {}
         out = func(*args, **kwargs)
         return self._count_flops(func._overloadpacket, out, args, kwargs)
```

and I observe these prints:

```
aten.conv2d.default
aten.add_.Tensor
aten.batch_norm.default
aten.conv2d.default
aten.add_.Tensor
aten.batch_norm.default
aten.add_.Tensor
aten.relu_.default
aten.conv2d.default
aten.add_.Tensor
aten.batch_norm.default
aten.relu_.default
aten.conv2d.default
aten.add_.Tensor
aten.batch_norm.default
aten.add_.Tensor
aten.relu_.default
aten.adaptive_avg_pool2d.default
aten.flatten.using_ints
aten.linear.default
aten.sum.default
Module      FLOP    % Total
--------  ------  ---------
Global         0         0%
```

To manually apply a decomposition even in the presence of inference mode, you do something like this:
`func.decompose(*args, **kwargs)` and then test if the result is NotImplemented or not.

Tag me as reviewer when you send the PR.

**Comment by akshat-sj** [comment]
If no one is currently working on this issue, I would like to take it as a first contribution.

**Comment by Julio-Cesar-Code** [comment]
@akshat-sj , are you working on this issue? I would like to take it on as my first contribution.

**Comment by akshat-sj** [comment]
@Julio-Cesar-Code yeah I was busy for a while but am still working on it.

**Comment by mattcattb** [comment]
has any progress been made on this issue? I would love to help contribute.

**Comment by StepinSilence** [comment]
@mattcattb I am afraid not. Don't know if @akshat-sj is still working on this issue.

**Comment by akshat-sj** [comment]
yeah I haven't been able to make considerable progress due to commitments, it is up for grabs.
